#Imports

In [ ]:
import pandas as pd
import numpy as np
import seaborn as sns
from nltk.tokenize import word_tokenize
from nltk.corpus import stopwords
import string
import matplotlib.pyplot as plt
from datetime import datetime, timedelta
import os
import re
try: 
    from BeautifulSoup import BeautifulSoup
except ImportError:
    from bs4 import BeautifulSoup

In [ ]:
pd.options.mode.chained_assignment = None  # default='warn'

# Reading the Data

In [ ]:
PATH = "/content/drive/MyDrive/data mining 1 /201709301651_masters_portal.csv"

In [ ]:
df = pd.read_csv(PATH)
result_df = pd.read_csv(PATH)
col = df.columns
program_count = len(df.index)

In [ ]:
result_df['keywords'] = ""

In [ ]:
all_keywords = [] 

In [ ]:
for c in col:
  nan = result_df[c].isna().sum()
  print(f'Column Name: {c}, Number of existing values: {len(result_df[c])-nan}, number of missing values: {nan}')

Column Name: country_name, Number of existing values: 60425, number of missing values: 0
Column Name: country_code, Number of existing values: 60425, number of missing values: 0
Column Name: university_name, Number of existing values: 60425, number of missing values: 0
Column Name: university_rank, Number of existing values: 34901, number of missing values: 25524
Column Name: program_name, Number of existing values: 60425, number of missing values: 0
Column Name: program_type, Number of existing values: 60425, number of missing values: 0
Column Name: deadline, Number of existing values: 29219, number of missing values: 31206
Column Name: duration, Number of existing values: 51656, number of missing values: 8769
Column Name: language, Number of existing values: 60404, number of missing values: 21
Column Name: tution_1_currency, Number of existing values: 53773, number of missing values: 6652
Column Name: tution_1_money, Number of existing values: 53773, number of missing values: 6652
Co

#First Column - Country Name
So far, we know there is no NaN value in this column.

In [ ]:
df['country_name'].unique()

array(['Armenia', 'Aruba', 'Australia', 'Austria', 'Azerbaijan',
       'Bahrain', 'Bangladesh', 'Barbados', 'Belgium',
       'Bosnia and Herzegovina', 'Brazil', 'Bulgaria', 'Canada', 'China',
       'Colombia', 'Croatia', 'Cyprus', 'Czech Republic', 'Denmark',
       'Egypt', 'Estonia', 'Ethiopia', 'Fiji', 'Finland', 'France',
       'Georgia', 'Germany', 'Greece', 'Grenada', 'Guam', 'Guatemala',
       'Guyana', 'Hong Kong', 'Hungary', 'Iceland', 'India', 'Indonesia',
       'Iran, Islamic Republic of', 'Ireland', 'Isle of Man', 'Israel',
       'Italy', 'Japan', 'Kazakhstan', 'Kenya', 'Kyrgyzstan', 'Latvia',
       'Lebanon', 'Liechtenstein', 'Lithuania', 'Luxembourg', 'Macao',
       'Macedonia, the former Yugoslav Republic of', 'Malaysia', 'Malta',
       'Monaco', 'Namibia', 'Netherlands', 'New Zealand', 'Nigeria',
       'Norway', 'Oman', 'Pakistan', 'Palestinian Territory, Occupied',
       'Philippines', 'Poland', 'Portugal', 'Qatar', 'Romania',
       'Russian Federation', '

In [ ]:
result_df

country_name country_code                             university_name  \
0          Armenia          ARM              American University of Armenia   
1          Armenia          ARM              American University of Armenia   
2          Armenia          ARM              American University of Armenia   
3          Armenia          ARM              American University of Armenia   
4          Armenia          ARM              American University of Armenia   
...            ...          ...                                         ...   
60420     Viet Nam          VNM         Engineering Institute of Technology   
60421     Viet Nam          VNM         Engineering Institute of Technology   
60422     Viet Nam          VNM         Engineering Institute of Technology   
60423     Viet Nam          VNM         Engineering Institute of Technology   
60424     Viet Nam          VNM  Vietnam National University of Agriculture   

       university_rank                                   program_name  \
0                  NaN                                      Economics   
1                  NaN    Political Science and International Affairs   
2                  NaN                        Business Administration   
3                  NaN               Computer and Information Science   
4                  NaN  Industrial Engineering and Systems Management   
...                ...                                            ...   
60420              NaN                             Electrical Systems   
60421              NaN                          Industrial Automation   
60422              NaN                  Renewable Energy Technologies   
60423              NaN                                         Safety   
60424              NaN                                Food Technology   

                   program_type              deadline   duration language  \
0                           MSc  2004-07-18T00:00:00Z        NaN  English   
1                        Master  2031-07-18T00:00:00Z  24 months  English   
2                           MBA  2004-07-18T00:00:00Z        NaN  English   
3                           MSc                   NaN  24 months  English   
4                          MEng  2031-07-18T00:00:00Z  24 months  English   
...                         ...                   ...        ...      ...   
60420                      MEng  2031-05-18T00:00:00Z        NaN  English   
60421                      MEng  2031-05-18T00:00:00Z  24 months  English   
60422  Postgraduate Certificate  2009-05-18T00:00:00Z   6 months  English   
60423                      MEng  2031-05-18T00:00:00Z        NaN  English   
60424                       MFT  2021-07-18T00:00:00Z  18 months  English   

      tution_1_currency  ...  tution_2_type tuition_price_specification  \
0                   EUR  ...         EU/EEA              Tuition (Year)   
1                   EUR  ...       National              Tuition (Year)   
2                   EUR  ...         EU/EEA              Tuition (Year)   
3                   EUR  ...       National              Tuition (Year)   
4                   EUR  ...       National              Tuition (Year)   
...                 ...  ...            ...                         ...   
60420               NaN  ...            NaN                         NaN   
60421               NaN  ...            NaN                         NaN   
60422               NaN  ...            NaN                         NaN   
60423               NaN  ...            NaN                         NaN   
60424               USD  ...       National    Tuition (Full programme)   

                  start_date  ielts_score  \
0        2018-09-01 00:00:00          6.5   
1        2018-08-22 00:00:00          6.5   
2        2018-09-01 00:00:00          6.5   
3                        NaN          6.5   
4        2018-08-22 00:00:00          6.5   
...                      ...          ...   
60420    2018-06-26 00:00:00          6

In [ ]:
for x in df['country_name']:
  f'in {x}'
  print([x])

Streaming output truncated to the last 5000 lines.
['United States']
['United States']
['United States']
['United States']
['United States']
['United States']
['United States']
['United States']
['United States']
['United States']
['United States']
['United States']
['United States']
['United States']
['United States']
['United States']
['United States']
['United States']
['United States']
['United States']
['United States']
['United States']
['United States']
['United States']
['United States']
['United States']
['United States']
['United States']
['United States']
['United States']
['United States']
['United States']
['United States']
['United States']
['United States']
['United States']
['United States']
['United States']
['United States']
['United States']
['United States']
['United States']
['United States']
['United States']
['United States']
['United States']
['United States']
['United States']
['United States']
['United States']
['United States']
['United States']
['United Stat

#Second Columns - University Name
So far, we know there is no NaN value in this column.

In [ ]:
university_names = df['university_name'].value_counts()

In [ ]:
# result_df = pd.concat([result_df, df['university_name']],axis=1)

#Third Column - University Rank
Numbers: 34901
NaNs: 25524

In [ ]:
# result_df = pd.concat([result_df, df['university_rank']],axis=1)

Some program dont have university rank while there are other program in same university that have a rank. So, we found them and filled them with the right number of rank.

In [ ]:
nan_rank_universities = {}
for i in range(program_count):
  uni_name = ""
  if pd.isnull(df.iloc[i]['university_rank']):
    uni_name = df['university_name'][i]
    if uni_name in nan_rank_universities:
      nan_rank_universities[uni_name][0] += 1
      nan_rank_universities[uni_name][1].append(i)
    else:
      nan_rank_universities[uni_name] = [1,[i]]

In [ ]:
for name in university_names.index:
  if name in nan_rank_universities:
    if university_names[name] > nan_rank_universities[name][0] :
      row_num = nan_rank_universities[name][1]
      for n in row_num:
        if result_df['university_name'][n-1] == name:
          result_df['university_rank'][n] = result_df['university_rank'][n-1]

In [ ]:
desc = result_df['university_rank'].describe()
print(desc)

count    34912.000000
mean       278.789671
std        202.151599
min          1.000000
25%        113.000000
50%        251.000000
75%        401.000000
max        801.000000
Name: university_rank, dtype: float64


Decided that replacing NaN with the mean would be good choice

In [ ]:
result_df['university_rank'].fillna(desc['mean'], inplace= True)

#4th Column - Program Name
No Nan value

In [ ]:
# result_df = pd.concat([result_df, df['program_name']],axis=1)

In [ ]:
df['program_name'].value_counts().sort_values(ascending=False)

Computer Science                                               435
Master of Business Administration                              387
Economics                                                      360
History                                                        359
Chemistry                                                      353
                                                              ... 
Education - Technology Concentration                             1
Elementary Education - Middle Level Education Concentration      1
Kinesiology - Sport Pedagogy Concentration                       1
Teaching - Community College Education                           1
Marine and Environmental Science                                 1
Name: program_name, Length: 25352, dtype: int64

#5th Column - Program Type
No NaN values

In [ ]:
df['program_type'].value_counts().sort_values(ascending=False)

MSc           24715
MA            13017
Master        10008
MBA            2677
MEd            1697
              ...  
MSGA              1
MGEOG             1
Foundation        1
MMR               1
MSAT              1
Name: program_type, Length: 502, dtype: int64

#6th Column - Deadline
Number of non-NaNs: 29219, number of NaNs: 31206

<hr />

In [ ]:
passed_deadlines = []

In [ ]:
invalid_deadlines = []

Logically, the "deadline" should be before the "start date" and the interval between them should be in months, not years. In the following piece of code, we tried to find the ones that don't satisfy the said conditions
Passed: any time before 2016

In [ ]:
# result_df['deadline'] = df['deadline']
result_df['start_date'] = df['start_date']

In [ ]:
def check_start_deadline_conditions(start, deadline, arr):
  if int(deadline.split('-')[0]) <= 2017:
    if 'You can apply until' in result_df['facts'][i]:
      deadline_date = result_df['facts'][i].split()[8]
      if '-' in deadline_date: # if deadline date was mentioned in facts
        result_df['deadline'][i] = deadline_date
      else:
        invalid_deadlines.append(i)
  else:
    if not pd.isnull(start):
      # 
      start = start.split()[0]
      if deadline < start: # checking if deadline is before start day
       result_df['deadline'][i] = deadline
      else:
        deadline_date = result_df['facts'][i].split()[8]
        if '-' in deadline_date: # if deadline date was mentioned in facts
          result_df['deadline'][i] = deadline_date
        else:
           invalid_deadlines.append(i)  

In [ ]:
for i in range(program_count):
 if not pd.isnull(result_df['deadline'][i]) :
   start = result_df['start_date'][i]
   deadline = result_df['deadline'][i].split('T')[0]
   check_start_deadline_conditions(start, deadline, i)

In [ ]:
for i in range(program_count):
  deadline = result_df['deadline'][i]
  if not pd.isnull(deadline) and 'T' in deadline:
    result_df['deadline'][i] = deadline.split('T')[0]

In [ ]:
interval_array = []
for i in range(program_count):
  if not pd.isnull(result_df['deadline'][i]) and not pd.isnull(result_df['start_date'][i]):
    if not i in invalid_deadlines:
      deadline = result_df['deadline'][i]
      start = result_df['start_date'][i].split()[0]
      interval_array.append((datetime.fromisoformat(start) - datetime.fromisoformat(deadline)).days)

In [ ]:
avg_interval = sum(interval_array)/len(interval_array)

In [ ]:
d = pd.DataFrame(columns=['deadline'])
d['deadline'] = result_df['deadline']
d.dropna(inplace= True)

In [ ]:
avg = str(np.array(d, dtype='datetime64[s]').view('i8').mean().astype('datetime64[s]'))
result_df['deadline'].fillna(avg.split('T')[0], inplace= True)

Handling NaN values

In [ ]:
def calculate_date(i):
 deadline_date = ""
 month = result_df['start_date'][i].split()[0].split('-')[1]
 year = result_df['start_date'][i].split()[0].split('-')[0]
 day = result_df['start_date'][i].split()[0].split('-')[2]
 month = int(month) - 3
 deadline_date = deadline_date.join([str(year),'-0',str(month),'-',str(day)])
 return deadline_date

#7th Column - duration
Number of non-NaNs: 51656, number of NaNs: 8769
<hr />

In [ ]:
df['duration'].describe()

count         51656
unique           80
top       24 months
freq          23336
Name: duration, dtype: object

In [ ]:
df['duration'].unique()

array([nan, '24 months', '12 months', '6 months', '18 months',
       '20 months', '21 months', '8 months', '36 months', '2 months',
       '15 months', '24', '48 months', '36', '72 months', '1 month',
       '30 months', '3 months', '17 months', '29 months', '21 days',
       '16 months', '84 months', '22 months', '23 months', '38 months',
       '25 months', '9 months', '70 days', '12', '60 months', '19 months',
       '4 months', '10 months', '14 months', '11 months', '32 months',
       '13 months', '26 months', '42 months', '27 months', '7 months',
       '5 months', '33 months', '28 months', '44 months', '90 months',
       '13', '66 months', '18 days', '24 days', '273 days', '29',
       '10 days', '54 months', '96 months', '3 days', '12 days', '60',
       '39 months', '40 months', '210 days', '238 days', '84 days', '48',
       '36 days', '35 months', '14 days', '34 months', '45 months',
       '7 days', '168 days', '98 days', '98 months', '196 days', '72',
       '41 months',

first we remove 'months'/'days' and convert all time to day

In [ ]:
# def convert_to_day(s):
type_list = []
for i, item in enumerate(result_df['duration']):
  if not pd.isnull(item):
    if isinstance(item, int):
      continue
    elif len(item) >= 1:
      if len(item.split()) > 1:
        num = int(item.split()[0])
        if item.split()[1] == 'months':
          result_df['duration'][i] = num * 30
        else:
          result_df['duration'][i] = num
      else:
        result_df['duration'][i] = int(item)

In [ ]:
print(result_df['duration'].min(),
    result_df['duration'].mean(),
    result_df['duration'].max())

1 608.4562103143875 21660


In [ ]:
for i, item in enumerate(result_df['duration']):
  if pd.isnull(item):
    result_df['duration'][i] = 608.4562103143875

In [ ]:
for i in range(program_count):
  if not pd.isnull(df.iloc[i]['facts']):
    if len(df['facts'][i].split("    "))>2:
      d_date = df['facts'][i].split("    ")[2]
      print(d_date.split()[0])
  else:
    print(f'NaN at {i}')

Streaming output truncated to the last 5000 lines.
2018-01-15
2018-01-07
2018-03-27
2017-12-31
Starting
2018-02-01
2018-02-01
2018-01-15
2017-12-01
2018-02-15
2018-02-01
2017-10-01
2017-12-15
2018-01-15
Time
2017-09-15
2018-01-15
Always
Duration
Time
2018-05-31
2018-02-18
Time
2018-02-27
2018-02-01
2018-02-01
Starting
Starting
2017-10-01
2018-01-10
2017-09-15
2018-04-15
2017-12-01
2018-01-15
2018-04-01
2018-02-01
2017-12-01
Starting
Starting
Starting
2018-03-01
2017-10-01
2017-10-15
2018-06-30
2017-10-15
2017-10-15
2017-10-15
2017-10-15
2017-10-15
2017-10-15
2017-10-15
2017-10-15
2017-10-15
2017-10-15
2017-11-15
2017-11-01
2017-11-15
2017-11-15
2017-11-15
2017-11-15
2017-11-15
2017-10-15
2017-10-15
2018-06-30
2018-03-01
2017-11-15
2017-12-20
2017-11-01
2017-12-15
Always
2018-02-15
2017-11-01
2017-11-15
2017-10-15
Always
Always
2018-02-01
2018-05-01
2018-05-01
2018-05-01
2018-05-01
2018-05-01
2018-05-01
Always
2018-05-01
2018-05-01
2018-05-01
2018-05-01
2017-12-01
2017-12-01
2017-12-01


In [ ]:
def group_by_values(l):
    count = {}
    for i in range(len(l)):
        item = l[i]
        if item in count.keys():
            count[item] += 1
        else:
            count[item] = 0
    return count

#8th Column - Language
Number of non-NaNs: 60404, number of NaNs: 21
<hr/>

In [ ]:
result_df['country_name'] = df['country_name']

In [ ]:
result_df['language'].describe()

count       60404
unique         75
top       English
freq        58864
Name: language, dtype: object

Replce NaN with the most reapeated language in the country

In [ ]:
def most_repeated_language(x):
  lang = result_df.loc[result_df['country_name'] == x]['language']
  # print(lang.value_counts().index[0])
  return lang.value_counts().index[0]

In [ ]:
for i, item in enumerate(result_df['language']):
  if pd.isnull(item):
    country_name = result_df['country_name'][i]
    result_df['language'][i] = most_repeated_language(country_name)

# 12th Column - Tuition Price Specification
Number of existing values: 53776, number of missing values: 6649
<hr/>

Change all to semester

1 semester = 4 month <br/>
1 Year = 2 semester <br/>
15 credit = 1 semester <br/>
1 Full Programme = depends on duration <br/>
4 Module = 1 semester <br/>
3 trimester = 2 semester <br/>
2 quarter = 1 semester <br/>

In [ ]:
result_df['tuition_price_specification'].value_counts().sort_values(ascending=False)

Tuition (Year)              32649
Tuition (Credit)             7661
Tuition (Full programme)     5416
Tuition (Semester)           4649
Tuition (Module)             3377
Tuition (Month)                13
Tuition (Trimester)             7
Tuition (Quarter)               4
Name: tuition_price_specification, dtype: int64

In [ ]:
result_df['tution_1_money'] = df['tution_1_money']

In [ ]:
def full_proframme_to_semester(duration, price):
  # duration is in days
  month = duration / 12
  return price / 3

In [ ]:
def convert_to_semester(price_str, price, duration):
  if price_str == 'Tuition (Year)':
    return price / 2
  elif price_str == 'Tuition (Credit)':
    return price*15
  elif price_str == 'Tuition (Full programme)':
    return full_proframme_to_semester(duration, price)
  elif price_str == 'Tuition (Module)' or price_str == 'Tuition (Month)':
    return price * 4
  elif price_str == 'Tuition (Trimester)':
    return price * 3 / 2
  elif price_str == 'Tuition (Quarter)':
    return price * 2
  else:
    return price

In [ ]:
for i in range(program_count):
  if not pd.isna(result_df['tuition_price_specification'][i]):
    if not pd.isna(result_df['tution_1_money'][i]):
      result_df['tution_1_money'][i] = convert_to_semester(result_df['tuition_price_specification'][i],
                                                      result_df['tution_1_money'][i], 
                                                      result_df['duration'][i]
                                                        )
      result_df['tuition_price_specification'][i] = 'Tuition (Semester)'

#9th Column - Tution 1 Currency
Number of non-NaNs: 53773, number of NaNs: 6652
<hr/>

In [ ]:
result_df['tution_1_currency'].value_counts().sort_values(ascending=False)

USD     25131
GBP     13726
EUR     10771
AUD      2421
CAD       928
Free      796
Name: tution_1_currency, dtype: int64

In [ ]:
description = result_df['tution_1_currency'].describe()
top_currency = description['top']
print(description)

count     53773
unique        6
top         USD
freq      25131
Name: tution_1_currency, dtype: object


In [ ]:
result_df['tution_1_currency'].fillna(top_currency, inplace= True)

# 10th Column - Tution 1 Money
Number of existing values: 53773, number of missing values: 6652
<hr/>

In [ ]:
description = result_df['tution_1_money'].describe()
print(description)

count     53773.000000
mean      10073.708701
std       15842.970344
min           0.000000
25%        5000.000000
50%        7800.000000
75%       12561.000000
max      618750.000000
Name: tution_1_money, dtype: float64


In [ ]:
result_df['tution_1_money'].fillna(description['mean'], inplace= True)

# 11th Column - Tution 1 Type
Number of existing values: 53773, number of missing values: 6652
<hr/>

In [ ]:
description = result_df['tution_1_type'].describe()
print(description)

count             53773
unique                4
top       International
freq              50980
Name: tution_1_type, dtype: object


In [ ]:
result_df['tution_1_type'].unique()

array(['International', nan, 'National', 'EU/EEA', 'In-State'],
      dtype=object)

In [ ]:
result_df['tution_1_type'].fillna(description['top'], inplace= True)

In [ ]:
result_df['tution_2_currency'].fillna(description['top'], inplace= True)

# 13th Column - Start Date
Number of existing values: 39099, number of missing values: 21326
<hr/>

In [ ]:
result_df['start_date'].describe()

count                     39093
unique                      319
top         2018-09-01 00:00:00
freq                       7658
Name: start_date, dtype: object

In [ ]:
def add_interval(finish_date, interval):
  finish_date_format = datetime.fromisoformat(finish_date)
  return str(finish_date_format + timedelta(days=interval))

In [ ]:
avg_start = add_interval(result_df['deadline'][i], avg_interval)
result_df['start_date'].fillna(avg_start.split()[0], inplace= True)

# 14th Column - Ielts Score
Number of existing values: 47357, number of missing values: 13068 
<hr/>

In [ ]:
desc = result_df['ielts_score'].describe()

In [ ]:
result_df['ielts_score'].fillna(desc['mean'], inplace= True)

# 15th Column - Structure
Number of existing values: 50523, number of missing values: 9902
<hr/>

# 16th Column - Academic Req 
Number of existing values: 59647, number of missing values: 778
<hr/>

When there is no \<li> or \<p> elements, information must be in section only.

In [ ]:
result_df['academic_req'] = df['academic_req']

In [ ]:
def get_gpa(s):
  gpa = 0
  if 'GPA' in s:
    if 'GPA of' in s:  
      gpa_index = s.split('GPA of')[1]
      if gpa_index.split()[0].isdigit():
        gpa = gpa_index.split()[0]
      elif 'better on a' in gpa_index:
        if len(gpa_index.split('better on a'))>1:
           gpa = 4.0
  return gpa

In [ ]:
def get_lang(i, x):
  lan = result_df['language'][i]
  
  if lan in x:
    return f'proficient in {lan}'

In [ ]:
def get_keywords(i, li_list):
  for i, x in enumerate(li_list):
    keyword_list = []
    x = x.text
    gpa = float(get_gpa(x))
    if gpa > 0:
      keyword_list.append(f'got gpa of {gpa}')
    if 'proficiency' in x:
      keyword_list.append(get_lang(i, x))
    if 'GRE' in x:
      keyword_list.append("took GRE exam")
  return keyword_list

Checking the ones with \<li>

In [ ]:
for i in range(program_count):
 
  s = result_df['academic_req'][i]
  
  if not pd.isna(s):
    parsed_html = BeautifulSoup(s)

    if parsed_html.find('li'):
      li_list = parsed_html.body.find_all('li')
      k_list = get_keywords(i, li_list)
      # if not k_list and not None in k_list:
      result_df['academic_req'][i] = k_list

\<p>

In [ ]:
def get_keywords_p(i, s):
  for i, x in enumerate(s):
    keyword_list = []
    x = x.text
    if 'GPA' in x:
      keyword_list.append(f'got gpa of 3.0')
    if 'proficiency' in x:
      lang = get_lang(i, x)
      if not lang is None:
        keyword_list.append(lang)
      if 'GRE' in x: 
        keyword_list.append("took GRE exam")
  return keyword_list

In [ ]:
for i in range(program_count):
  s = result_df['academic_req'][i]
  
  if not isinstance(s, list) and not pd.isna(s):
    
    parsed_html = BeautifulSoup(s)
    
    if parsed_html.find('p'):
      p_list = parsed_html.body.find_all('p')
      l = get_keywords_p(i, p_list)
      result_df['academic_req'][i] = l

In [ ]:
def get_gpa_section(s):
   gpa = 0
   if 'GPA' in s:
    if 'GPA of' in s:
      gpa = (s.split('GPA of')[1].split()[0])
    else:
      next = s.split('GPA')[1].split()[0]
      if next.isdigit():
        gpa = next
      else:
        if '/' in next:
         gpa = next.split('/')[0]
    if not type(gpa) is int and ')' in gpa:
      gpa = gpa.split(')')[0]
    if not type(gpa) is int and gpa[-1] == '.':
      gpa = gpa[:-1]
   return gpa

In [ ]:
def section_get_keywords(s):
  keyword_list = []
  s = s.replace('Academic Requirements','')
 
  gpa = get_gpa_section(s)
  if gpa!=0 :
    if gpa.isnumeric():
      keyword_list.append(f'got gpa of {gpa}')
    elif '.' in gpa:
      keyword_list.append(f'got gpa of {gpa}')
  if 'proficiency' in x:
      lang = get_lang(i, x)
      if not lang is None:
        keyword_list.append(lang)
  if 'GRE' in x: 
    keyword_list.append("took GRE exam")

In [ ]:
for i in range(program_count):
  s = result_df['academic_req'][i]
  if not isinstance(s, list) and not pd.isna(s):
    parsed_html = BeautifulSoup(s)
    if not parsed_html.find('li') and not parsed_html.find('p'):
      section = parsed_html.body.find('section').text
      result_df['academic_req'][i] = section_get_keywords(section)

# Keyword Extraction

In [ ]:
all_keywords = []

In [ ]:
all_keywords

[['a MSc degree',
  'in Economics',
  'at American University of Armenia',
  'of rank 278.78967117323555',
  'in Armenia',
  'affordable for me',
  "It's in English",
  'got ielts of 6.5',
  'interested in Quantitative Methods for Economists (Mathematical Economics),  Econometrics 1 (Economic Statistics),  Microeconomic Theory,  Macroeconomic Theory,  Econometrics 2, ',
  'took GRE exam'],
 ['a Master degree',
  'in Political Science and International Affairs',
  'at American University of Armenia',
  'of rank 278.78967117323555',
  'in Armenia',
  'affordable for me',
  "It's in English",
  'got ielts of 6.5'],
 ['a MBA degree',
  'in Business Administration',
  'at American University of Armenia',
  'of rank 278.78967117323555',
  'in Armenia',
  'affordable for me',
  "It's in English",
  'got ielts of 6.5',
  'interested in Managers with practical knowledge of accounting,  finance,  marketing and other functional areas essential for success in a dynamic and globalized business envi

## 1. Add program type and program name

In [ ]:
for i in result_df.index:
  p_type = result_df['program_type'][i]
  all_keywords.append([f'a {p_type} degree'])

In [ ]:
for i in result_df.index:
  p_name = result_df['program_name'][i]
  all_keywords[i].append(f'in {p_name}')

## 2. Add university name ad its rank

In [ ]:
for i in result_df.index:
  uni = result_df['university_name'][i]
  all_keywords[i].append(f'at {uni}')

In [ ]:
for i in result_df.index:
  rank = result_df['university_rank'][i]
  all_keywords[i].append(f'of rank {rank}')

## 3. Add country name

In [ ]:
for i in result_df.index:
  country = result_df['country_name'][i]
  all_keywords[i].append(f'in {country}')

## 4. Add tution

In [ ]:
for i in result_df.index:
  currency = result_df['tution_1_currency'][i]
  if currency == 'Free':
    all_keywords[i].append(f'Free')

In [ ]:
avg = result_df['tution_1_money'].describe()['mean']
for i in result_df.index:
  fee = result_df['tution_1_money'][i]
  if 0 < fee <= avg :
    all_keywords[i].append(f'affordable for me')

## 5. Languaes

In [ ]:
for i in result_df.index:
  lang = result_df['language'][i]
  all_langauges = ''.join(l for l in lang)
  all_langauges = all_langauges.replace(',',' and')
  all_keywords[i].append(f'It\'s in {all_langauges}')

## Ielts score

In [ ]:
for i in result_df.index:
  score = "{:.1f}".format(result_df['ielts_score'][i])
  all_keywords[i].append(f'got ielts of {score}')

## Structure

In [ ]:
for i in result_df.index:
  s = result_df['structure'][i]
  if not pd.isna(s):
   s = s.replace('\'',"").replace('[',"").replace(']',"")
   st = s.split(',')[:5]
   struct =  ''.join(f'{l}, ' for l in st)
   all_keywords[i].append(f'interested in {struct}')

## Academic Requirement

In [ ]:
for i in result_df.index:
  req = result_df['academic_req'][i]
  if not type(req) is float and req and not None in req:
    all_keywords[i].append(''.join(r for r in req))
  

In [ ]:
result_df['all_keywords'] = all_keywords

<hr/>

In [ ]:
temp = result_df

In [ ]:
for i in invalid_deadlines:
  temp = temp.drop(i)

In [ ]:
temp.to_csv('final.csv')

In [ ]:
keywords = result_df['all_keywords'][10]
input_string = ','.join(r for r in keywords)


In [ ]:
x = 'write a statement of purpose for me, I am applying for'
x + input_string 

"write a statement of purpose for me, I am applying fora Postgraduate Certificate degree,in Asset and Maintenance Management,at CQUniversity Australia,of rank 401.0,in Australia,affordable for me,It's in English,got ielts of 6.0,interested in Introduction to Asset and Maintenance Management,  Asset Management Systems,  Maintenance Strategies,  Maintenance Organisation, "